### ISSUE TITLE
Latest circe-java8 no more defines Defaults time enc/decoders preventing upgrading circe to 0.10.0 

### ISSUE THREAD

**Issue by sledorze** [twilio/guardrail#103]
Regression with circe-java8 0.10.0 (0.9.3 does not exhibit that issue).

These are used here:

https://github.com/twilio/guardrail/blob/4e45d6b3c9df6f5b6ab9306fb086ea1d7eeaf1a9/modules/codegen/src/main/scala/com/twilio/guardrail/generators/CirceProtocolGenerator.scala#L340-L355

But no more defined:
https://github.com/circe/circe/blob/master/modules/java8/src/main/scala/io/circe/java8/time/JavaTimeDecoders.scala#L184

The methods have been renamed.

I'm new to Guardrail / Circe / etc.. wondering how / if I could help with it..

**Comment by blast-hardcheese** [comment]
So, to start, welcome! Thank you for the report!

Looking at these implicits do:

```scala
implicit def decodeInstant: Decoder[Instant] = j8time.decodeInstant.or(decodeLong.map(Instant.ofEpochMilli))
implicit def decodeLocalDate: Decoder[LocalDate] = j8time.decodeLocalDateDefault.or(decodeInstant.map(_.atZone(ZoneOffset.UTC).toLocalDate))
implicit def decodeOffsetDateTime: Decoder[OffsetDateTime] = j8time.decodeOffsetDateTimeDefault.or(decodeInstant.map(_.atZone(ZoneOffset.UTC).toOffsetDateTime))

// Mirror
implicit val encodeInstant = j8time.encodeInstant
implicit val encodeLocalDateDefault = j8time.encodeLocalDateDefault
implicit val encodeLocalDateTimeDefault = j8time.encodeLocalDateTimeDefault
implicit val encodeLocalTimeDefault = j8time.encodeLocalTimeDefault
implicit val encodeOffsetDateTimeDefault = j8time.encodeOffsetDateTimeDefault
implicit val encodeZonedDateTimeDefault = j8time.encodeZonedDateTimeDefault
```

`decodeInstant`:
- Attempts to decode an `Instant` using whatever format is default to circe's java8 time formatter.
- If that fails, we attempt to parse a long and treat it as Epoch Millis (:exclamation:)

`decodeLocalDate`:
- Attempt to decode a `LocalDate` using whatever format is default to circe's java8 time formatter.
- If that fails, we attempt to decode an instant and assume it's UTC (:exclamation:)

`decodeOffsetDateTime`:
- Attempts to decode an OffsetDateTime using whatever format is default to circe's java8 time formatter.
- If that fails, we attempt to decode an instant and assume it's UTC (:exclamation:)

It would be good to break the fallback handlers out into explicit vendor extensions if these are actually desirable. Instead of `type: string` and `format: date-time` secretly handling milliseconds, it would be better to have `type: integer` and `format: int64` with a vendor extension that specifies that the values are actually UTC milliseconds (or nanoseconds, or seconds).

Once an answer is determined for how to handle these fallback scenarios, describing a new type that represents the transition between `Long` and `OffsetDateTime` could look like this:

```scala
case class EpochMillis(value: OffsetDateTime)
implicit val encodeMillis: Encoder[EpochMillis] =
  Encoder[Long]
    .contramap[OffsetDateTime](_.toInstant.toEpochMilli)
    .contramap[EpochMillis](_.value)
implicit val decodeMillis: Decoder[EpochMillis] =
  Decoder[Long]
    .map(Instant.ofEpochMilli _)
    .map(_.atZone(ZoneOffset.UTC).toOffsetDateTime)
    .map(EpochMillis.apply _)
```

This would not be exposed to the user directly, similar to how [`respond` members](https://github.com/dpblh/guardrail/blob/a64a04df309224967bfe27c9ebbb550f6282d557/src/test/scala/tests/generators/akkaHttp/AkkaHttpServerTest.scala#L181-L183) or [file upload parts](https://github.com/twilio/guardrail/blob/8818189/modules/codegen/src/main/scala/com/twilio/guardrail/generators/AkkaHttpServerGenerator.scala#L613) work.

I don't have answers about exactly how this should work, from a user perspective though, or from what already exists in the broader OpenAPI community though. What do you think?

**Comment by sledorze** [comment]
@blast-hardcheese I think there's two issues here - correct me if I'm wrong.

1) A change in circe-java8 which prevent compiling.
2) Fallback handling to millis which is very implicit and should be corrected.

1) is a Show stopper for upgrading circe.
2) is a Bug that should be resolved (a breaking change if someone secretly relies on it).

Am I correct?
If so I propose to split the issue.

**Comment by blast-hardcheese** [comment]
I believe #135 has resolved the first point, #113 is still a concern

**Comment by blast-hardcheese** [comment]
please reopen if necessary
